// @technote-space/github-action-version-helper/dist/utils/version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNextVersion = exports.getNextVersionLevel = exports.whatBump = exports.getCurrentVersion = void 0;
const github_action_helper_1 = require("@technote-space/github-action-helper");
const commit_1 = require("./commit");
const misc_1 = require("./misc");
const constant_1 = require("../constant");
const getCurrentVersion = async (helper) => helper.getLastTag();
exports.getCurrentVersion = getCurrentVersion;
const whatBump = (minorUpdateCommitTypes, commits) => {
    if (commits.filter(commit => commit.notes.length).length) {
        return 'major';
    }
    if (minorUpdateCommitTypes.length && commits.filter(commit => commit.type && minorUpdateCommitTypes.includes(commit.type)).length) {
        return 'minor';
    }
    return 'patch';
};
exports.whatBump = whatBump;
const getNextVersionLevel = (minorUpdateCommitTypes, commits) => constant_1.VERSION_BUMP[(0, exports.whatBump)(minorUpdateCommitTypes, commits)];
exports.getNextVersionLevel = getNextVersionLevel;
const getNextVersion = async (minorUpdateCommitTypes, excludeMessages, breakingChangeNotes, helper, octokit, context, logger) => {
    const commits = await (0, commit_1.getCommits)(minorUpdateCommitTypes, excludeMessages, breakingChangeNotes, octokit, context);
    (0, misc_1.log)(logger => logger.startProcess('Target commits:'), logger);
    (0, misc_1.log)(() => console.log(commits
        .filter(item => item.notes.length || item.type)
        .map(item => ({
        type: item.type,
        message: item.message,
        notes: item.notes,
        sha: item.sha,
    }))), logger);
    (0, misc_1.log)(logger => logger.endProcess(), logger);
    const current = await (0, exports.getCurrentVersion)(helper);
    (0, misc_1.log)(logger => logger.info('Current version: %s', current), logger);
    const next = github_action_helper_1.Utils.generateNewVersion(current, (0, exports.getNextVersionLevel)(minorUpdateCommitTypes, commits));
    (0, misc_1.log)(logger => logger.info('Next version: %s', next), logger);
    return next;
};
exports.getNextVersion = getNextVersion;
