// @technote-space/github-action-version-helper/dist/utils/version.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNextVersion = exports.getNextVersionLevel = exports.whatBump = exports.getCurrentVersion = void 0;
const github_action_helper_1 = require("@technote-space/github-action-helper");
const commit_1 = require("./commit");
const misc_1 = require("./misc");
const constant_1 = require("../constant");
exports.getCurrentVersion = (helper) => __awaiter(void 0, void 0, void 0, function* () { return helper.getLastTag(); });
exports.whatBump = (minorUpdateCommitTypes, commits) => {
    if (commits.filter(commit => commit.notes.length).length) {
        return 'major';
    }
    if (minorUpdateCommitTypes.length && commits.filter(commit => commit.type && minorUpdateCommitTypes.includes(commit.type)).length) {
        return 'minor';
    }
    return 'patch';
};
exports.getNextVersionLevel = (minorUpdateCommitTypes, commits) => constant_1.VERSION_BUMP[exports.whatBump(minorUpdateCommitTypes, commits)];
exports.getNextVersion = (minorUpdateCommitTypes, excludeMessages, breakingChangeNotes, helper, octokit, context, logger) => __awaiter(void 0, void 0, void 0, function* () {
    const commits = yield commit_1.getCommits(minorUpdateCommitTypes, excludeMessages, breakingChangeNotes, octokit, context);
    misc_1.log(logger => logger.startProcess('Target commits:'), logger);
    misc_1.log(() => console.log(commits
        .filter(item => item.notes.length || item.type)
        .map(item => ({
        type: item.type,
        message: item.message,
        notes: item.notes,
        sha: item.sha,
    }))), logger);
    misc_1.log(logger => logger.endProcess(), logger);
    const current = yield exports.getCurrentVersion(helper);
    misc_1.log(logger => logger.info('Current version: %s', current), logger);
    const next = github_action_helper_1.Utils.generateNewVersion(current, exports.getNextVersionLevel(minorUpdateCommitTypes, commits));
    misc_1.log(logger => logger.info('Next version: %s', next), logger);
    return next;
});
